### PR TITLE
refactor(common): refactor validation pipe

### PR DESCRIPTION
**Pull request by jbpionnier**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: packages/common/pipes/validation.pipe.ts [hunk 1]
```diff
@@ -16,13 +16,12 @@ let classTransformer: any = {};
 
 @Injectable()
 export class ValidationPipe implements PipeTransform<any> {
-  protected isTransformEnabled: boolean;
-  protected isDetailedOutputDisabled?: boolean;
-  protected validatorOptions: ValidatorOptions;
+  protected readonly isTransformEnabled: boolean;
+  protected readonly isDetailedOutputDisabled?: boolean;
+  protected readonly validatorOptions: ValidatorOptions;
```
Comment (kamilmysliwiec) [inline packages/common/pipes/validation.pipe.ts]: What if someone decides to extend `ValidationPipe` and change these values asynchronously (depending on some async resources, remote config service, whatever)? In such case, TypeScript would report errors about changing `readonly` values outside of the `constructor`.

File: packages/common/pipes/validation.pipe.ts [hunk 2]
```diff
@@ -59,11 +58,11 @@ export class ValidationPipe implements PipeTransform<any> {
     if (type === 'custom') {
       return false;
     }
-    const types = [String, Boolean, Number, Array, Object];
-    return !types.some(t => metatype === t) && !isNil(metatype);
+    const types = [String, Boolean, Number, Array, Object] as ReadonlyArray<Type<any>>;
```
Comment (marcus-sa) [inline packages/common/pipes/validation.pipe.ts]: There's absolutely 0 reason to type the array, as TS by default does that, which means the array is typed as `any[]`

File: packages/common/pipes/validation.pipe.ts [hunk 3]
```diff
@@ -59,11 +58,11 @@ export class ValidationPipe implements PipeTransform<any> {
...
+    return !types.some((t: Type<any>) => metatype === t);
```
Comment (marcus-sa) [inline packages/common/pipes/validation.pipe.ts]: The `Type<T>` interface is only used to type class references, which means that it isn't a valid use case.

File: packages/common/pipes/validation.pipe.ts [hunk 4]
```diff
@@ -16,13 +16,12 @@ let classTransformer: any = {};
...
-  protected isDetailedOutputDisabled?: boolean;
-  protected validatorOptions: ValidatorOptions;
+  protected readonly isTransformEnabled: boolean;
+  protected readonly isDetailedOutputDisabled?: boolean;
+  protected readonly validatorOptions: ValidatorOptions;
 
   constructor(@Optional() options?: ValidationPipeOptions) {
-    options = options || {};
-    const { transform, disableErrorMessages, ...validatorOptions } = options;
+    const { transform, disableErrorMessages, ...validatorOptions }: ValidationPipeOptions  = options || {};
```
Comment (marcus-sa) [inline packages/common/pipes/validation.pipe.ts]: This is NOT right.
You shouldn't use the OR operator to default options to an empty object in this situation, and you shouldn't type the object twice, instead it should be like this.

```ts
constructor(@Optional() options: ValidationPipeOptions = {}) {
  const { transform, disableErrorMessages, ...validatorOptions } = options;
```

**Comment by jbpionnier** [comment]
ok... 
it's PR is not necessary.
I close

**Pull request closed without merging**